lib/get-client.js

```diff
@@ -1,14 +1,123 @@
 const url = require('url');
-const GitHubApi = require('@octokit/rest');
+const {memoize} = require('lodash');
+const Octokit = require('@octokit/rest');
+const pRetry = require('p-retry');
+const pThrottle = require('p-throttle');
 
-module.exports = (githubToken, githubUrl, githubApiPathPrefix) => {
+/**
+ * Default exponential backoff configuration for retries.
+ */
+const DEFAULT_RETRY = {retries: 3, factor: 2, minTimeout: 1000};
+
+/**
+ * Rate limit per API endpoints.
+ *
+ * See {@link https://developer.github.com/v3/search/#rate-limit|Search API rate limit}.
+ * See {@link https://developer.github.com/v3/#rate-limiting|Rate limiting}.
+ */
+const RATE_LIMITS = {
+  search: [30, 60 * 1000],
+  core: [5000, 60 * 60 * 1000],
+};
+
+/**
+ * Global rate limit to prevent abuse.
+ *
+ * See {@link https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits|Dealing with abuse rate limits}
+ */
+const GLOBAL_RATE_LIMIT = [1, 1000];
```

Comment:
we only need to wait a second for mutating requsets 

> If you're making a large number of POST, PATCH, PUT, or DELETE requests for a single user or client ID, wait at least one second between each request.

`GET` AND `HEAD` requests are balanced across many servers so no need to limit these. But maybe that's overkill for semantic-release to make the differentiation

Reply:
It would be quite easy to implement in the current code. I just have to include or not the `globalThrottler`. The problem is to determine which client functions correspond to a `POST`, `PATCH`, `PUT`, or `DELETE`. Is there a way to know that?

Reply:
Or I can add a `throttle` argument that would be passed in each call. So whenever I call a client function in other part of the code and pass this argument:
```js
await github.repos.get(repo, false);
await github.repos.createRelease(release, true);
```

But 90% of the call we make are:
- `search` which has a very low rate limit anyway
- `repos.createRelease`, `repos.uploadAsset`, `issues.create`,  `issues.edit` and `issues.createComment` which all are `POST`, `PATCH`, `PUT` or DELETE

Reply:
yeah let's leave as is